cmd/crank/beta/convert/pipelinecomposition/converter.go

```diff
@@ -147,15 +147,98 @@ func processFunctionInput(input *Input) *runtime.RawExtension {
 		"apiVersion":  "pt.fn.crossplane.io/v1beta1",
 		"kind":        "Resources",
 		"environment": processedInput.Environment.DeepCopy(),
-		"patchSets":   processedInput.PatchSets,
-		"resources":   processedInput.Resources,
+		"patchSets":   MigratePatchPolicy(processedInput.PatchSets),
+		"resources":   MigratePatchPolicyInResources(processedInput.Resources),
 	}
 
 	return &runtime.RawExtension{
 		Object: &unstructured.Unstructured{Object: inputType},
 	}
 }
 
+// MigratePatchPolicyInResources patches resources.
+func MigratePatchPolicyInResources(resources []v1.ComposedTemplate) []ComposedTemplate {
+	composedTemplates := []ComposedTemplate{}
+
+	for _, resource := range resources {
+		composedTemplate := ComposedTemplate{}
+		composedTemplate.ComposedTemplate = resource
+		composedTemplate.Patches = patchPatches(resource.Patches)
+
+		composedTemplate.ComposedTemplate.Patches = nil
+		composedTemplates = append(composedTemplates, composedTemplate)
+	}
+	return composedTemplates
+}
+
+// MigratePatchPolicy migrates mergeoptions.
+func MigratePatchPolicy(patchset []v1.PatchSet) []PatchSet {
+	newPatchSets := []PatchSet{}
+
+	for _, patchSet := range patchset {
+		newpatchset := PatchSet{}
+		newpatchset.Name = patchSet.Name
+		newpatchset.Patch = patchPatches(patchSet.Patches)
+
+		newPatchSets = append(newPatchSets, newpatchset)
+	}
+
+	return newPatchSets
+}
+
+func patchPatches(patches []v1.Patch) []Patch {
+	newPatches := []Patch{}
+
+	for _, patch := range patches {
+		newpatch := Patch{}
+		newpatch.Patch = patch
+
+		if patch.Policy != nil && patch.Policy.MergeOptions != nil {
+			newpatch.Policy = patchPolicy(patch.Policy)
+			newpatch.Patch.Policy = nil
+		}
+
+		newPatches = append(newPatches, newpatch)
+	}
+
+	return newPatches
+}
+
+func patchPolicy(policy *v1.PatchPolicy) *PatchPolicy {
+	if policy.FromFieldPath == nil {
+		return nil
+	}
+
+	pp := &PatchPolicy{
+		FromFieldPath: policy.FromFieldPath,
+	}
+
+	mo := policy.MergeOptions
```

Comment:
I don't think this logic is 100% correct for the conversion rules we want to enable. Let's review the following resources to guide the conversion logic this function should capture:

* https://github.com/crossplane-contrib/function-patch-and-transform/issues/101#issuecomment-2024141807
* https://github.com/crossplane-contrib/function-patch-and-transform/issues/101#issuecomment-2024155035
* https://github.com/crossplane-contrib/function-patch-and-transform/?tab=readme-ov-file#differences-from-the-native-implementation


Issues:
* `fromFieldPath` probably doesn't matter, we should be able to ignore that, right?
* it doesn't handle the case of `Replace`
* it is only considering values being `nil`, it's not handling values being set explicitly to `false`

Here is a logic table that captures the conversion rules we need to codify in this function. I've kept the case ordering @negz used in his linked comments above for consistency:

| # | `MergeOptions` | `appendSlice`   | `keepMapValues` | `policy.toFieldPath` |
| -----------  | -----------  | ------------- | -----------  | ------------- |
| 1 | `nil`  | N/A | N/A | `Replace` |
| 2 | `non-nil` | `nil` or `false` | `true` | `MergeObjects` | 
| 3 | `non-nil` | `true` | `nil` or `false` | `ForceMergeObjectsAppendArrays` |
| 4 | `non-nil` | `nil` or `false` |  `nil` or `false` |  `ForceMergeObjects`   |
| 5 | `non-nil` | `true` | `true` | `MergeObjectsAppendArrays` |

Because this function is only considering `nil` values, the results it produces doesn't quite match up with what I expect. For example, if the function takes in `{appendSlice: true, keepMapValues: false}` (case 3 above), it produces the incorrect result of `MergeObjectsAppendArrays`.

What do you think of this analysis, is it missing anything? If you agree with it, then please update this function to cover these cases and produce these results 🙏

Reply:
Thanks for the detailed summary on this, I have handled all the cases now.

Reply:
Awesome @padlar, the functionality does indeed look correct now, and the new unit tests are thorough with good coverage for the cases! Nice work, thank you for getting this functionality to something we can rely on 😉 

One thing I realized is that it's probably better to just set the converted policy to `nil` when nothing was specified in the original manifest, so we aren't adding config (bloat) that wasn't actually in the original manifest. That means that case 1 above should just be `nil` meaning function-p-and-t will use its default of `Replace` ([code link](https://github.com/crossplane-contrib/function-patch-and-transform/blob/main/input/v1beta1/resources_patches.go#L90)). I edited the above table to reflect that.

Another thing I was thinking of is the readability and future maintainability of this logic. We took awhile together to actually figure out what the logic is supposed to be, so I was thinking it would be wise to make this function super readable for our future selves to look at it and immediately know what it's doing.

I took a stab at a very readable approach in https://github.com/jbw976/crossplane/tree/merge-options-conversion, specifically https://github.com/jbw976/crossplane/commit/e56e06678e664d3e2b2c517a79605a5d81559804. What do you think of that style? I think I like it because it feels more natural and quickly readable. What do you think about adopting it? Feel free to just copy that code, no need to add me as a committer or anything 🤓

Reply:
Good catch on setting the converting policy to `nil`. And totally agree on making the conversion logic simple to follow and maintain :) I have copy pasted your code now :)